### ISSUE TITLE
Release 2023-03 follow-up

### ISSUE THREAD

**Issue by stonebig** [winpython/winpython#1244]
target: ~~august~~ mid-september 2023

wanted:
- [x] Jupyterlab-4
- [x] PEP-514 compatibility
- [x] Python-3.11.5,
- [x] bokeh-3.2
- [x] numba-0.58 final, numpy-1.25 (in pursuite of [ data-apis standard](https://data-apis.org/array-api/latest/changelog.html) )

focus:
- Jupyterlab-4
- simplify to standard: PEP-514

exploratory:
- LargeLangageModel teaching per Andrew Ng:
   - 1 hour : [ChatGPT Prompt Engineering for Developers](https://learn.deeplearning.ai/chatgpt-prompt-eng)
   - 1 hour: [LangChain for LLM Application Development](https://learn.deeplearning.ai/langchain)
- LargeLangageModel on various WinPython problems: sans-Qt interface, build, clean-up, semi-automated DataScience ...
- a local LLM that could work on a gpu-less PC:
   - https://simonwillison.net/2023/May/5/mpt-7b/
   - https://github.com/ggerganov/llama.cpp
   - https://github.com/rustformers/llm
   - ... llm-0.7 from Simon Willison

probably postponed:
- practice & build LLMs toolbox from factual learning examples on WinPython
- multi-interpreters , nogil ... seems 3.13  
- simplify winpython build
- jedi-0.19 big speed-up is with spyder-5.5...  next cycle
- micromamba study

helping to push the Python envelop:
- ubiquity: Jupyterlab-4, VSCode-Jupyter, Jupyterlite, Python-in-Excel
- speed in general: Cython-3, mypyc, latest Python with multi-interpreters, nogil-3.13
- speed in datascience:
   - the move to "array API standard" and to "pyarrow" latest numpy compatible with numba, latest Pandas
   - rust packages accelerators : duckdb, polars, pydantic, ...
- ... ml & llm on local

**Comment by RoyiAvital** [comment]
You have NumPy 1.25.1 with MKL: https://github.com/cgohlke/numpy-mkl-wheels

**Comment by RoyiAvital** [comment]
There is issue with `umap`. For some reason the packages include `umap` while they should include `umap-learn`.
See https://umap-learn.readthedocs.io.

**Comment by stonebig** [comment]
audit:
- spyder-5.4.4
- cartopy-0.22.0 official wheel

**Comment by stonebig** [comment]
WinPython 2023-03 b1 (2023-08-06)

 MD5                             | SHA-1                                    | SHA-256                                                          | Binary                            | Size               | blake2b-256                                                     
---------------------------------|------------------------------------------|------------------------------------------------------------------|-----------------------------------|--------------------|-----------------------------------------------------------------
a944d0ae194832d1d0b971de230c1d82 | 17dc86afa078122864fcda3bbfa1d0d4f07ac581 | 4edcf65f1bdf8757a0a4126a342ba47742995c1d9dbe7ecc94da9f3b1a9491c0 | Winpython64-3.11.4.1dotb1.exe     |   24 628 301 Bytes | de85fe2d0da12cbfa949845cda6b98e10f036223609b7b085b29feadf1476c49
4f01263813d806a59d285687bc8af07f | 9c8673b85f012c90d21ce45eccd91fa3d09d50a6 | 2b2bb1c3dc725d236c5f08100361514987ef8316c9236aed970e489ebef05567 | Winpython64-3.11.4.1b1.exe        |  783 717 435 Bytes | 5036e13fa8546ffd1d4c7f278547e3ea71a8d5df57874fc7a0ce5e7a42d3a79f

**Comment by stonebig** [comment]
WinPython 2023-03 b2 (2023-08-1x)

Changes from WinPython 2023-03 b1:
- Infrastructure:
   - python-3.12.0rc1 , pip-23.2.1
- New Packages:
   - llm-0.7 
- Upgraded Packages: 
   - pydantic-2.1.1 (moves to rust)
   - fastapi-101

Areas of particular interest for testers:
- fastapi

**Comment by stonebig** [comment]
WinPython 2023-03 b3 (2023-08-20)

Changes from WinPython 2023-03 b2:
- Infrastructure:
   - PEP-514 : when you register per `wppm --register ` or per graphical wppm menu,
- New Packages:
- Upgraded Packages: 
   - numba-0.58.0rc1,  numpy-1.25.2
   - polars-0.18.15

Areas of particular interest for testers:
- integration via PEP-0514, even for '.dot' version
- numpy-1.25.2 

next build:
- python-3.11.5 
- numba-0.58 final

binaries:
- https://github.com/winpython/winpython/releases/tag/7.0.20230820
- https://sourceforge.net/projects/winpython/files/WinPython_3.11/3.11.4.1/betas/


 MD5                             | SHA-1                                    | SHA-256                                                          | Binary                            | Size               | blake2b-256                                                     
---------------------------------|------------------------------------------|------------------------------------------------------------------|-----------------------------------|--------------------|-----------------------------------------------------------------
d2bd632b2a6aaaf87979bfefb8eb3e81 | 132b7608c13abe1d3b757c8cbfc638ad587b48ee | 87a1d4b55f826e120657544adcbef2c886885486bdfe76701adc40f3fa5ce5c6 | Winpython64-3.11.4.1dotb3.exe     |   24 679 788 Bytes | 7eac83457c2620d6f50cb2c8ce2fbbe9af552eb73cf95fdeb7d8593f4d9ed2b9
a09057b9ca987b8bc775aaae99352543 | fc18e082e85fa24eb20e38e97038212643f97e4e | 1fa82d9ded80f0d0bea11989c8a924072e678c88079bd7a994672d6ed7e68307 | Winpython64-3.11.4.1b3.exe        |  803 001 083 Bytes | 7b9c3b797643927cd0c9f84c02b6f1b8d6aa3b06b95c4c8dc5f89580697093ea
80aa40b39ee852dd361405e22b0746ba | e1196075db22d871b08059508e04b141d36f6fee | 8fc90e0347b12d4e851aff62a73687ca50b8d82f5572c1ba0c25673d026d6381 | Winpython64-3.12.0.0dotrc1b.exe   |   23 858 335 Bytes | 96e7a9d75cdb682bd2e3f967a7071bcf26f6131730ae7a90c7a2ac0a4509f221

**Comment by stonebig** [comment]
> You have NumPy 1.25.1 with MKL: https://github.com/cgohlke/numpy-mkl-wheels

Intel does efforts to merge some work in standard Numpy: https://numpy.org/doc/stable/release/1.25.0-notes.html#faster-np-sort-on-avx-512-enabled-processors

**Comment by RoyiAvital** [comment]
Still not in the same order of magnitude as using MKL.

**Comment by stonebig** [comment]
> Still not in the same order of magnitude as using MKL.

ok. trying a spin to see. let there be light, in the fiber.

**Comment by stonebig** [comment]
WinPython 2023-03 b4 (2023-08-2x)

**Comment by fleming79** [comment]
The batch file "scripts/winvsode.bat" has changed meaning the VS code launcher doesn't work as I'm used to with loading from /t/ folder. 
This feature was quite handy. Any chance of restoring it?

```
@echo off
rem launcher for VScode
call "%~dp0env_for_icons.bat" %*
rem cd/D "%WINPYWORKDIR1%"
if exist "%WINPYDIR%\..\t\vscode\code.exe" (
    "%WINPYDIR%\..\t\vscode\code.exe" %*
) else (
if exist "%LOCALAPPDATA%\Programs\Microsoft VS Code\code.exe" (
    "%LOCALAPPDATA%\Programs\Microsoft VS Code\code.exe"  %*
) else (
    "code.exe" %*
))

```

**Comment by stonebig** [comment]
Ok, will testore the Feature at next build

**Comment by firai** [comment]
3.11.5 lands a day after you release a build

**Comment by stonebig** [comment]
Thoughts on Python-in-Excel (executing the Python Code in Azure):
- you'll have to pay extra to execute Python Cycles on azure (normal / un-avoidable)
- so Python-in-Excel will be a usable advantage when it can run also on a local instance of Python, with no extra
- in the mean time, it's very good publicity for Python.

**Comment by stonebig** [comment]
> 3.11.5 lands a day after you release a build

and so another beta with this and numba-rc2 to happen

**Comment by stonebig** [comment]
WinPython 2023-03 b5 (2023-09-0x)

Changes from WinPython 2023-03 b4:
- Infrastructure:
   - Python-3.11.5
   - node-18.17.1
- New Packages:
- Upgraded Packages: 
   - bokeh-3.2.2, panel-1.2.2
   - Cython-3.0.2, spyder-5.4.5
- Remove packages:
- Differences by version
   - mkl version has numpy-MKL added and Pytorch directly related eco-system removed.  

Areas of particular interest for testers:
- bokeh, Panel
- cython

next build:
- numba-0.58 rc2 or final

**Comment by hiccup7** [comment]
The MKL versions are listed here: https://github.com/cgohlke/numpy-mkl-wheels
But there are newer versions available for non-MKL SciPy, NumPy and NumExpr.
I propose separate lists of WinPython packages for MKL and non-MKL, and use the newer versions for non-MKL.

**Comment by RoyiAvital** [comment]
I wish `pip` will allow, somehow, a solution like `conda`.  
Otherwise, To make WinPython sustainable, it is probably better to leave `MKL` out unless @stonebig embrace `micromamba`.

**Comment by hiccup7** [comment]
Numba 0.58.0rc2 is available:
https://pypi.org/project/numba/0.58.0rc2/
https://github.com/numba/numba/issues/9160

**Comment by stonebig** [comment]
WinPython 2023-03 b6 (2023-09-1x)

Changes from WinPython 2023-03 b5:

- Infrastructure:
- New Packages:
- Upgraded Packages:
  - numba-0.58.0rc2
  - jupyterlab-4.0.6
  - geopandas-0.14.0
- Removed packages:
- Differences by version
  - mkl version has numpy-MKL added and Pytorch directly related eco-system removed.


Areas of particular interest for testers:
- numba-0.58.0rc2

next build: (rc)
- numba-0.58  final 

binaries:

**Comment by stonebig** [comment]
WinPython 2023-03 rc (2023-09-2x)

Changes from WinPython 2023-03 b6:

- Infrastructure:
- New Packages:
- Upgraded Packages:
  - numba-0.58.0
  - matplotlib-3.8.0, pandas-2.1.1, scikit-learn-1.31.1
  - panel-1.2.3
- Removed packages:
- Differences by version
  - mkl version has numpy-MKL added and Pytorch directly related eco-system removed.


Areas of particular interest for testers:
- nbdime (git comparison) remains as of now broken for Jupyterlab-4.0
- no major miss found

**Comment by RoyiAvital** [comment]
I'd wait for [Numba `0.58.1` for support for NumPy `1.26`](https://github.com/numba/numba/issues/9201).
Otherwise it is outdated out of the gate.

**Comment by hiccup7** [comment]
Winpython64-3.11.5.0rc.exe has NumPy 1.25.2.  NumPy 1.26 is a big change in the way NumPy is built, so it has a big risk of bugs.  I suggest waiting until WinPython release 2023-04 to update to NumPy 1.26 and Numba 0.58.1.

Winpython64-3.11.5.0rc.exe has SciPy v1.11.1.  If rc2 is needed for a serious bug, I suggest an update to SciPy v1.11.2, where 90% of the changes affect Python 3.11.

**Comment by stonebig** [comment]
> Winpython64-3.11.5.0rc.exe has NumPy 1.25.2. NumPy 1.26 is a big change in the way NumPy is built, so it has a big risk of bugs. I suggest waiting until WinPython release 2023-04 to update to NumPy 1.26 and Numba 0.58.1.

no numpy-1.26 this release, as  numba-0.58.0 was already awaited since 3 weeks to get numpy-1.25 compatibility.

> Winpython64-3.11.5.0rc.exe has SciPy v1.11.1. If rc2 is needed for a serious bug, I suggest an update to SciPy v1.11.2, where 90% of the changes affect Python 3.11.

scipy-1.11.2 is a little miss ... but seems not to contain a major bugfix in it as far as I can see

**Comment by firai** [comment]
nbdime 4.0.0a0 has preliminary support of JL 4. There are, of course, a number of bugs to be fixed.

**Comment by stonebig** [comment]
> scipy-1.11.2 is faster on modern hardware due to a new openblas version.

hum, I didn't notice that line, but would not be a bug

> nbdime 4.0.0a0 has preliminary support of JL 4. There are, of course, a number of bugs to be fixed.

it was indeed looking still pretty bad, I suppose they try to be ok for october 9th 
![image](https://github.com/winpython/winpython/assets/4312421/b20deec5-0c21-4252-81ec-f14bdcb03fc9)

**Comment by stonebig** [comment]
Hello,

As of September 28th, no major complains occured, so rc binaries are now re-tagged as final without any change.

Binaries:
- https://github.com/winpython/winpython/releases/tag/7.0.20230928
- https://sourceforge.net/projects/winpython/files/WinPython_3.11/3.11.5.0

 MD5                             | SHA-1                                    | SHA-256                                                          | Binary                            | Size               | blake2b-256                                                     
---------------------------------|------------------------------------------|------------------------------------------------------------------|-----------------------------------|--------------------|-----------------------------------------------------------------
e7c9bbe0480296192a087475214d82d0 | c1d70b0602357614f993975ebff149cb8db4d56b | 9399d24addcfa14450b408c1988b0350f4e9d58ab2ae9a2445b9dab6484f4fae | Winpython64-3.11.5.0dot.exe       |   25 139 984 Bytes | 368c68f07f5717ec69523f0d0ca3d3de97394c489baa74301a2ad498a767c43e
c6a63eca7bb0ed59612781cb508ce4a7 | 0a1f13fac46a9457ca26144cedc2b6d981ba5d9b | 5b053362706b2cb9561ebd064570813dafc99598b7c7d9473e7361313b820178 | Winpython64-3.11.5.0.exe          |  790 793 444 Bytes | 21a9014a64ed7f4ab8668599d744ee54a14893e62f6ad9b5a394ce7b96040d6b
9fe0fd44b0f33c4d23739245c5b491af | 24d47348cf722c79ba354b854733c4e2d808cc83 | 0b2037660064589dd73a58ac17f44d9e657cfc5fcea92cec9223185243a0a3bc | Winpython64-3.11.5.0mkl.exe       |  789 143 611 Bytes | c09c0831480e2da584509fa8e2a2ea9a98d87e84eebf5eb8f9061e8ec9b82e31

**Comment by RoyiAvital** [comment]
I thought you were waiting for `numba 0.58.1`.

**Comment by stonebig** [comment]
Numba has to adapt to next numpy-1.26, and numpy-1.26+ eco-system  has to stablise, while adapting to cpython-3.12
...  so it was looking like about 4 weeks away
